File: premarket_alert/scanner.py
from __future__ import annotations

import logging

from .config import Config
from .models import Mover

log = logging.getLogger(__name__)


def passes_filters(mover: Mover, config: Config) -> tuple[bool, str]:
    """Voldoet deze mover aan de drempels? Tweede waarde is de reden bij afwijzing."""
    if mover.change_pct < config.threshold_pct:
        return False, f"stijging {mover.change_pct:.1f}% < {config.threshold_pct:.1f}%"
    if mover.price < config.min_price:
        return False, f"koers {mover.price:.2f} < {config.min_price:.2f}"
    if config.max_price > 0 and mover.price > config.max_price:
        return False, f"koers {mover.price:.2f} > {config.max_price:.2f}"
    if mover.volume < config.min_premarket_volume:
        return False, f"volume {mover.volume} < {config.min_premarket_volume}"
    if mover.dollar_volume < config.min_dollar_volume:
        return False, f"omzet ${mover.dollar_volume:,.0f} < ${config.min_dollar_volume:,.0f}"
    if config.exchanges and mover.exchange and mover.exchange.upper() not in {e.upper() for e in config.exchanges}:
        return False, f"beurs {mover.exchange} niet in selectie"
    if config.instrument_types and mover.instrument_type not in config.instrument_types:
        return False, f"type {mover.instrument_type} niet in selectie"
    return True, ""


def select(movers: list[Mover], config: Config) -> list[Mover]:
    """Filter op de drempels en sorteer van hardste stijger naar minste."""
    hits: list[Mover] = []
    for mover in movers:
        ok, reason = passes_filters(mover, config)
        if ok:
            hits.append(mover)
        elif mover.change_pct >= config.threshold_pct:
            log.debug("%s afgewezen: %s", mover.symbol, reason)

    hits.sort(key=lambda m: m.change_pct, reverse=True)
    return hits[: config.max_results]

File: worker/src/scanner.ts
import type { Config } from "./config.ts";
import { dollarVolume, type Mover } from "./models.ts";

/** Voldoet deze mover aan de drempels? Bij afwijzing de reden, anders null. */
export function rejectReason(m: Mover, cfg: Config): string | null {
  if (m.changePct < cfg.thresholdPct) return `stijging ${m.changePct.toFixed(1)}% < ${cfg.thresholdPct.toFixed(1)}%`;
  if (m.price < cfg.minPrice) return `koers ${m.price.toFixed(2)} < ${cfg.minPrice.toFixed(2)}`;
  if (cfg.maxPrice > 0 && m.price > cfg.maxPrice) return `koers ${m.price.toFixed(2)} > ${cfg.maxPrice.toFixed(2)}`;
  if (m.volume < cfg.minPremarketVolume) return `volume ${m.volume} < ${cfg.minPremarketVolume}`;
  if (dollarVolume(m) < cfg.minDollarVolume) {
    return `omzet $${Math.round(dollarVolume(m))} < $${Math.round(cfg.minDollarVolume)}`;
  }
  const exchanges = cfg.exchanges.map((e) => e.toUpperCase());
  if (exchanges.length && m.exchange && !exchanges.includes(m.exchange.toUpperCase())) {
    return `beurs ${m.exchange} niet in selectie`;
  }
  if (cfg.instrumentTypes.length && !cfg.instrumentTypes.includes(m.instrumentType)) {
    return `type ${m.instrumentType} niet in selectie`;
  }
  return null;
}

/** Filter op de drempels en sorteer van hardste stijger naar minste. */
export function select(movers: Mover[], cfg: Config, debug: (msg: string) => void = () => {}): Mover[] {
  const hits: Mover[] = [];
  for (const m of movers) {
    const reason = rejectReason(m, cfg);
    if (reason === null) hits.push(m);
    else if (m.changePct >= cfg.thresholdPct) debug(`${m.symbol} afgewezen: ${reason}`);
  }
  hits.sort((a, b) => b.changePct - a.changePct);
  return hits.slice(0, cfg.maxResults);
}
